doc/developer-notes.md

```diff
@@ -723,6 +723,47 @@ General Bitcoin Core
   - *Explanation*: If the test suite is to be updated for a change, this has to
     be done first.
 
+Logging
+-------
+
+The macros `LogInfo`, `LogDebug`, `LogTrace`, `LogWarning` and `LogError` are available for
+logging messages. They should be used as follows:
+
+- `LogDebug(BCLog::CATEGORY, fmt, params...)` is what you want
+  most of the time, and it should be used for log messages that are
+  useful for debugging and can reasonably be enabled on a production
+  system (that has sufficient free storage space). They will be logged
+  if the program is started with `-debug=category` or `-debug=1`.
+  Note that `LogPrint(BCLog::CATEGORY, fmt, params...)` is a deprecated
+  alias for `LogDebug`.
+
+- `LogInfo(fmt, params...)` should only be used rarely, eg for startup
```

Comment:
https://github.com/bitcoin/bitcoin/pull/28318/commits/f7ce5ac08c669ac763e275bb7c82dcfb2b1b6c33

It's somewhat wonky that e.g. `LogInfo()` doesn't take a category, but e.g. `LogDebug` does, and in later commits we actually remove category information from info-equivalent log statements, which seems backwards to me. Ideally, `LogInfo` would _optionally_ take a category. But maybe this can be done later?

Reply:
> Ideally, LogInfo would optionally take a category.

I had that intuition at first too, but then changed my mind after reading [ajtowns's comment](https://github.com/bitcoin/bitcoin/pull/28318#issuecomment-1702318968) stating:

> My reasoning here is that the usefulness of the "category" isn't in telling us what bit of code triggered the log message (we have -logsourcelocations for that which is much more precise, and you can generally just grep for the message anyway), but in categorising log messages the user can disable. It's useful for the user to see [net] because it tells them they can use -debug=net to enable it; it's useful for the user to see [net:trace] because it tells them they also need to enable tracing to see that. It's useful to see [warning] or [error] because those are things they probably want to take action on, and having a consistent tag means they can grep for them more easily. It's not useful to see [net:info] because that doesn't tell them anything they can action, it's just noise that makes it harder to read the logs.
> 
> So I contend that approach is consistent, it's just consistent over a different (in my opinion more valuable) attribute; namely that categories are solely for conditional logging that the user can disable on a per-subsystem basis.

Given that LogInfo is unconditional, what would be your use case for adding a log category?

Reply:
> Given that LogInfo is unconditional, what would be your use case for adding a log category?

Simply that "info = unconditional" might not always be the case (and is an unusual convention to have hardcoded), and a user could rightly want to see only WARNING+ logs as well as, say, `net` or validation-related INFOs.

Reply:
I think ajtowns addressed that quite well at the bottom of [this comment](https://github.com/bitcoin/bitcoin/pull/28318#issuecomment-1702318968): 

> I don't think that's a sensible concern: unconditional logging should be rare enough to not need disabling (eg, connecting to new outbounds, new headers, and new tips for about 400 entries a day), and at that point it's better not to disable them as that gives the user a pretty clear indication their node is still working right. Even for users that only care about errors/warnings, having the UpdateTip context at least is likely useful for giving some clue what state the node was in when it encountered the problem.

Seeing warning/error-logs only is easy enough to do with grep, and info logging should be infrequent enough to not take up significant system resources.

Reply:
@stickies-v it's not about system resources; it's about allow the end user to articulate what they see. Not a big deal either way, but I think our current scheme is a little wonky if you compare it to more widely deployed systems e.g. the python logging module or log4j.

Reply:
> it's about allowing the end user to articulate what they see

which is trivial to do with `grep`? My point about system resources is that theoretically we could always log _everything_ but that would use an unreasonable amount of system resources, hence the nuance.

Reply:
> which is trivial to do with `grep`?

This doesn't seem trivial if some log messages have categories and some don't. Like there is no way anymore to only see a list of all the messages only from wallet, or all the messages only from net_processing. I don't think this is terrible, but it does seem weird, and I wonder if any log systems work like this where they go out of their way to define categories, but then choose not to include categories in the most important messages.

Reply:
> This doesn't seem trivial if some log messages have categories and some don't.

If you want all log messages to have categories for easier grepping (or for importing into a database), there's the `-loglevelalways=1` option introduced in this PR. 

> Like there is no way anymore to only see a list of all the messages only from wallet, 

If you want to focus on messages from some section of the code, then that's what the `-logsourcelocations=1` option is for.

Your wording suggests that you think this is a regression. It's not: we've traditionally used `LogPrintf()` for non-debug messages which also doesn't allow selecting only messages from the wallet, and does not add any consistent tagging.

> I don't think this is terrible, but it does seem weird, and I wonder if any log systems work like this where they go out of their way to define categories, but then choose not to include categories in the most important messages.

I don't think it is actually terribly weird compared to log4j: the equivalent there would be setting up a single global logger, that does info/warn/error levels and always logs all messages, and a bunch of subsystem-specific loggers that do debug/trace levels and that default to off. I think it'd be roughly the same with python: have a global logger for info/warn/error (`logging.warning("boo")`), and child loggers for subsystems for debug/trace (`logging.getLogger('net').debug('did something')`).

I'm not really sure what else to say -- I feel like the arguments stickies-v quoted above were pretty clear, and the result is both simple and useful, even if it seems "weird" or surprising.

Reply:
To clarify, I don't think that anyone is debating that this changeset is an improvement over the status quo. The ask (at least from my end) would be to unify the `Log*` interfaces and allow the optional specification of a category. That way, when migrating existing logging statements over, we're not throwing away category information just because something is INFO. And we support users who want to see WARN only most of the time, but care about, say, `net` INFO messages. That doesn't seem like a crazy thing to ask for.

Reply:
> Again, no opinion on this. I think either solution is fine. PRs welcome, I guess?

Thanks, I started playing around with a change that should make it easier to set the category without being verbose, and also start relying less on a global logging instance, which should be useful for kernel applications and tests: 262923c5b37f62a925a2c5611318855009bfb241. Will see how this looks and probably open a PR.

Reply:
> Will see how this looks and probably open a PR

Opened #29256 with these changes. Code changes should be complete but there are some test failures so it is a draft.